src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+#ifdef _WIN32
+#ifndef FD_SETSIZE
+#define FD_SETSIZE 2048  // Set max fds - only for select() fallback
+#endif
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include <atomic>
+#include "tinycthread.h"
+#include "later.h"
+#include "callback_registry_table.h"
+
+#if R_VERSION < R_Version(4, 5, 0)
+inline SEXP R_mkClosure(SEXP formals, SEXP body, SEXP env) {
+  SEXP fun = Rf_allocSExp(CLOSXP);
+  SET_FORMALS(fun, formals);
+  SET_BODY(fun, body);
+  SET_CLOENV(fun, env);
+  return fun;
+}
+#endif
+
+extern CallbackRegistryTable callbackRegistryTable;
+extern SEXP later_fdcancel;
+extern SEXP later_invisibleSymbol;
+
+#define LATER_INTERVAL 1024
+
+#ifdef _WIN32
+#define POLL_FUNC WSAPoll
+#else
+#define POLL_FUNC poll
+#endif
+
+typedef struct ThreadArgs_s {
+  std::shared_ptr<std::atomic<bool>> flag;
+  std::unique_ptr<std::vector<int>> fds;
+  SEXP callback;
+  double timeout;
+  int rfds, wfds, efds;
+  int num_fds;
+  int loop;
+} ThreadArgs;
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  const bool flag = args->flag->load();
+  args->flag->store(true);
+  if (!flag) {
+    std::memcpy((void *) DATAPTR_RO(CADR(args->callback)), args->fds->data(), args->num_fds * sizeof(int));
+    Rf_eval(args->callback, R_GlobalEnv);
+  }
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: if necessary to add method for HANDLES on Windows. Would be different code to SOCKETs.
+// TODO: implement re-usable background thread.
+static int wait_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  int *fds = args->fds->data();
```

Comment:
Is there a specific reason you'd rather work with a raw pointer here than `std::vector`? I/we generally try to stay in modern C++ land where possible.

Reply:
It's what I *want* to do given I know it's safe here, but in no way necessary! So I respect your point and reverted the use in [3626011](https://github.com/r-lib/later/pull/190/commits/3626011dd598e1de61b3652cb1be549c1074cc90).